libraries/libopaemem/hash_map.c

```diff
@@ -0,0 +1,291 @@
+			       void (*key_cleanup)(void *key, void *context),
+			       void (*value_cleanup)(void *value, void *context))
+{
+	if (!hm || !key_hash || !key_compare) {
+		ERR("NULL pointer(s)");
+		return FPGA_INVALID_PARAM;
+	}
+
+	memset(hm, 0, sizeof(*hm));
+
+	hm->buckets = (opae_hash_map_item **)
+			opae_calloc(num_buckets,
+				    sizeof(opae_hash_map_item *));
+	if (!hm->buckets) {
+		ERR("calloc() failed");
+		return FPGA_NO_MEMORY;
+	}
+
+	hm->num_buckets = num_buckets;
+	hm->hash_seed = hash_seed;
+	hm->key_hash = key_hash;
+	hm->key_compare = key_compare;
+	hm->key_cleanup = key_cleanup;
+	hm->value_cleanup = value_cleanup;
+
+	return FPGA_OK;
+}
+
+STATIC opae_hash_map_item *
+opae_hash_map_alloc_item(void *key, void *value)
+{
+	opae_hash_map_item *item;
+	item = (opae_hash_map_item *)
+		opae_malloc(sizeof(opae_hash_map_item));
+	if (item) {
+		item->key = key;
+		item->value = value;
+		item->next = NULL;
+	}
+	return item;
+}
+
+fpga_result opae_hash_map_add(opae_hash_map *hm,
+			      void *key,
+			      void *value)
+{
+	uint32_t key_hash;
+	opae_hash_map_item *item;
+	opae_hash_map_item *list;
+
+	if (!hm) {
+		ERR("NULL pointer");
+		return FPGA_INVALID_PARAM;
+	}
+
+	item = opae_hash_map_alloc_item(key, value);
+	if (!item) {
+		ERR("malloc() failed");
+		return FPGA_NO_MEMORY;
+	}
+
+	key_hash = hm->key_hash(hm->num_buckets,
+				hm->hash_seed,
+				key);
+
+	list = hm->buckets[key_hash];
+	if (!list) {
+		// Bucket was empty. Fill it.
+		hm->buckets[key_hash] = item;
+		return FPGA_OK;
+	}
+
+	// Bucket not empty. Do we have a key collision?
+	while (list) {
+		if (!hm->key_compare(key, list->key)) {
+			// Key collision.
+			opae_free(item);
+			if (hm->value_cleanup)
+				hm->value_cleanup(list->value, hm->cleanup_context);
+			list->value = value; // Replace value only.
+			return FPGA_OK;
+		}
+		list = list->next;
+	}
+
+	// No key collision. Add item to the list.
+	item->next = hm->buckets[key_hash];
+	hm->buckets[key_hash] = item;
+
+	return FPGA_OK;
+}
+
+fpga_result opae_hash_map_find(opae_hash_map *hm,
+			       void *key,
+			       void **value)
+{
+	uint32_t key_hash;
+	opae_hash_map_item *list;
+
+	if (!hm) {
+		ERR("NULL pointer");
+		return FPGA_INVALID_PARAM;
+	}
+
+	key_hash = hm->key_hash(hm->num_buckets,
+				hm->hash_seed,
+				key);
+
+	list = hm->buckets[key_hash];
+
+	while (list) {
+		if (!hm->key_compare(key, list->key)) {
+			if (value)
+				*value = list->value;
+			return FPGA_OK;
+		}
+		list = list->next;
+	}
+
+	return FPGA_NOT_FOUND;
+}
+
+fpga_result opae_hash_map_remove(opae_hash_map *hm,
+				 void *key)
+{
+	uint32_t key_hash;
+	opae_hash_map_item *prev = NULL;
+	opae_hash_map_item *list;
+
+	if (!hm) {
+		ERR("NULL pointer");
+		return FPGA_INVALID_PARAM;
+	}
+
+	key_hash = hm->key_hash(hm->num_buckets,
+				hm->hash_seed,
+				key);
+
+	list = hm->buckets[key_hash];
+	if (!list)
+		return FPGA_NOT_FOUND;
+
+	while (list) {
+		if (!hm->key_compare(key, list->key))
```

Comment:
Calling a function through a pointer is slow in a critical routine, both here and in opae_hash_map_find(). I understand the value of setting a function to generalize hash comparison.

How about a compromise? It makes the code more complicated, but you could check whether hm->key_compare == opae_u64_key_hash. If it is, call opae_u64_key_hash directly instead of through a pointer. The branch predictor will handle that case correctly -- much easier than predicting a call through a pointer. You could even mark opae_u64_key_hash inline and avoid the jump inside the loop entirely.

Reply:
Implemented this optimization both for key_hash and key_compare.